File: 3. Longest Substring Without Repeating Characters.cpp
#include<iostream>
#include<unordered_map>
using namespace std;


class Solution {
    public:
        int lengthOfLongestSubstring(string s) {

            int i =0;
            int j = 0;

            int maxLen = 0;

            unordered_set<int> mySet;

            while(j<s.length()){

                while(mySet.find(s[j]) != mySet.end()){
                    mySet.erase(mySet.find(s[i]));
                    i++;
                }

                mySet.insert(s[j]);
                maxLen = max(maxLen,j-i+1);

                j++;

            }

            return maxLen;

        }
};




int main(){

    string input = "abba";

    Solution solution;
    int answer = solution.lengthOfLongestSubstring(input);
    
    cout << "Answer is: " << answer << endl;

}


    

File: 3. Longest Substring Without Repeating Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        if len(s) == 0: return 0
        
        seen = set()

        l = r = result = 0

        while r < len(s):
            
            while s[r] in seen:
                seen.remove(s[l])
                l += 1
            
            result = max(result,r-l)
            seen.add(s[r])
            r += 1

        return result + 1 
            